### ISSUE TITLE
Linking against cc_library that itself has system dependencies fails

### ISSUE THREAD

**Issue by guibou** [tweag/rules_haskell#537]
I have the following lib `lz4` imported in my `WORKSPACE`:

```
nixpkgs_package(
    name = "lz4",
    build_file_content = """
package(default_visibility = ["//visibility:public"])

cc_library (
    name = "lib3",
    srcs = ["lib/liblz4.so.1"],
    testonly = 1,
)
""",
    repository = "@nixpkgs",
)
```

As a reference, I have the following c file `LZ4Main.c':

```
extern int LZ4_decompress_safe(char*, char*, int, int);


int main()
{
	LZ4_decompress_safe(0, 0, 0, 0);
}
```

And the following haskell file `LZ4Main.hs`:

```
{-# LANGUAGE OverloadedStrings #-}

import Data.Word ( Word8)
import Foreign.C.Types
import Foreign.Ptr (Ptr)

-- | Decompress a bytestring.
foreign import ccall unsafe "lz4.h LZ4_decompress_safe"
  c_LZ4_decompress_safe :: Ptr CChar -- ^ source
                        -> Ptr Word8 -- ^ destination
                        -> CInt      -- ^ source size
                        -> CInt      -- ^ max decompressed size
                        -> IO CInt   -- ^ decompression result

main = do
  print =<< c_LZ4_decompress_safe undefined undefined undefined undefined
```

Here is my bazel file:

```
load(
    "@io_tweag_rules_haskell//haskell:haskell.bzl",
    "haskell_binary",
)

cc_binary(
    name = "cc_youpi",
    deps = ["@lz4//:lib3"],
    srcs = ["LZ4Main.c"],
    testonly=1,
    )

haskell_binary(
    name = "youpi",
    deps = ["@hackage//:base", "@lz4//:lib3"],
    srcs = ["LZ4Main.hs"],
    testonly=1,
    )
```

The `cc_binary` target builds and run, but the `haskell_binary` fails:

```
[nix-shell:~/tweag/rules_haskell]$ bazel run //tests/lz4:cc_youpi 
INFO: Invocation ID: 311cc325-3f6a-4241-ac60-4d6fab30880a
DEBUG: /home/guillaume/.cache/bazel/_bazel_guillaume/80819d925a91d7b49ed839165b536ee7/external/bazel_skylib/lib.bzl:30:1: WARNING: lib.bzl is deprecated and will go away in the future, please directly load the bzl file(s) of the module(s) needed as it is more efficient.
INFO: Analysed target //tests/lz4:cc_youpi (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //tests/lz4:cc_youpi up-to-date:
  bazel-bin/tests/lz4/cc_youpi
INFO: Elapsed time: 0.099s, Critical Path: 0.00s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action

[nix-shell:~/tweag/rules_haskell]$ bazel run //tests/lz4:youpi 
INFO: Invocation ID: edba54c3-9f3a-433e-aacf-8fd7b5107aa4
DEBUG: /home/guillaume/.cache/bazel/_bazel_guillaume/80819d925a91d7b49ed839165b536ee7/external/bazel_skylib/lib.bzl:30:1: WARNING: lib.bzl is deprecated and will go away in the future, please directly load the bzl file(s) of the module(s) needed as it is more efficient.
INFO: Analysed target //tests/lz4:youpi (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
ERROR: /home/guillaume/tweag/rules_haskell/tests/lz4/BUILD.bazel:14:1: error executing shell command: '/nix/store/9n9mh7kpz5d20w23mn1r9ig72zxxvnf0-bash/bin/bash -c ${1+"$@"} $(< bazel-out/k8-fastbuild/bin/tests/lz4/_obj/youpi.dynamic.manifest)  bazel-out/host/bin/tests/visible-binaries/ghc -pgma baz...' failed (Exit 1) bash failed: error executing command /nix/store/9n9mh7kpz5d20w23mn1r9ig72zxxvnf0-bash/bin/bash -c '${1+"$@"} $(< bazel-out/k8-fastbuild/bin/tests/lz4/_obj/youpi.dynamic.manifest)' '' bazel-out/host/bin/tests/visible-binaries/ghc -pgma ... (remaining 41 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
bazel-out/k8-fastbuild/bin/tests/lz4/_obj/youpi/Main.dyn_o:s3v7_info: error: undefined reference to 'LZ4_decompress_safe'
collect2: error: ld returned 1 exit status
`cc' failed in phase `Linker'. (Exit code: 1)
Target //tests/lz4:youpi failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 0.340s, Critical Path: 0.25s
INFO: 0 processes.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
```

This may be a regression somewhere (I'll investigate this today) because it was working well before in my codebase.

**Comment by mboes** [comment]
Bug has been there since at least September 1st. Unlikely to be a regression.

**Comment by mboes** [comment]
Also note that you should be using `haskell_cc_import` here, as in the following test: https://github.com/tweag/rules_haskell/blob/master/tests/binary-with-sysdeps/BUILD. Not that this fixes the test case.

**Comment by mboes** [comment]
I'm not understanding the use case. Is there are reason you are not using `cc_import` for the system library? The [documentation for rules_haskell](http://api.haskell.build/haskell/cc.html#haskell_cc_import) says that cc_import should be replaced with haskell_cc_import (temporary hack due to upstream limitations). If you `haskell_cc_import` the original library directly this all works.

**Comment by guibou** [comment]
Could you share your solution?

**Comment by guibou** [comment]
> The documentation for rules_haskell says that cc_import should be replaced with haskell_cc_import (temporary hack due to upstream limitations)

The documentation do not details what are the upstream limitations and I always assumed that it was only the lake of `strip_include_prefix`.

`cc_import` does not work with shared library on which the soname is different from the basename, see https://github.com/bazelbuild/bazel/issues/1534

**Comment by guibou** [comment]
```
 bazel run //tests/binary-with-sysdeps:binary-with-sysdeps 
INFO: Invocation ID: 3e50cc5a-4ddf-4c9a-a645-7f309103aba6
DEBUG: /home/guillaume/.cache/bazel/_bazel_guillaume/80819d925a91d7b49ed839165b536ee7/external/bazel_skylib/lib.bzl:30:1: WARNING: lib.bzl is deprecated and will go away in the future, please directly load the bzl file(s) of the module(s) needed as it is more efficient.
INFO: Analysed target //tests/binary-with-sysdeps:binary-with-sysdeps (27 packages loaded, 590 targets configured).
INFO: Found 1 target...
Target //tests/binary-with-sysdeps:binary-with-sysdeps up-to-date:
  bazel-bin/tests/binary-with-sysdeps/binary-with-sysdeps
INFO: Elapsed time: 6.013s, Critical Path: 0.79s
INFO: 15 processes: 15 processwrapper-sandbox.
INFO: Build completed successfully, 18 total actions
INFO: Build completed successfully, 18 total actions
/home/guillaume/.cache/bazel/_bazel_guillaume/80819d925a91d7b49ed839165b536ee7/execroot/io_tweag_rules_haskell/bazel-out/k8-fastbuild/bin/tests/binary-with-sysdeps/binary-with-sysdeps: error while loading shared libraries: libz.so.1: cannot open shared object file: No such file or directory
```

The already existing example is not working too.

**Comment by guibou** [comment]
Confirmed regression:

```
4a9a3245a7162f40ae739ae7e148fd32219ab4a8 is the first bad commit
commit 4a9a3245a7162f40ae739ae7e148fd32219ab4a8
Author: Philip Patsch <philip.patsch@tweag.io>
Date:   Wed Sep 5 16:31:55 2018 +0200

    link.bzl: Make rpaths refer to the runfiles directory
    
    The rpaths as they were constructed went all the way up to the bazel
    cache directory and then used the full path to reference dynamic
    libraries.
    
    This is conceptually wrong, since bazel targets only see their
    `runfiles` directory at runtime, which is in practice enforced by
    `bazel run`, which sandboxes the executed commands (at least on
    Linux).
    
    Instead of using the full path from the toplevel cache directory, we
    use the `short_path` of the file to point to, from the runfiles
    directory.

:040000 040000 dc127771936ab60003e6c324606c553441dd046a 83a86ea63d59e1b5e90d1f0fd3b95df6f1539844 M	haskell
bisect run success
```

Bisection script:

```
#!/usr/bin/env bash
nix-shell --command 'bazel run //tests/binary-with-sysdeps'
```

**Comment by Profpatsch** [comment]
I’m working on it.

**Comment by Profpatsch** [comment]
Actually, we should revert that commit for now and I’ll try to fix it without giving unsuspecting people even more linker errors.

As @guibou found out, this breaks most tests if you also run the binaries instead of just building them, which I was unaware of at the time.

**Comment by Profpatsch** [comment]
Temporarily fixed by the revert in #550.
